ni_measurement_plugin_sdk_service/_internal/parameter/encoder.py

```diff
@@ -0,0 +1,222 @@
+# enums and default values
+from enum import Enum
+from typing import Any, Dict, Sequence
+from uuid import uuid4
+
+from google.protobuf import descriptor_pb2, descriptor_pool, message_factory
+from google.protobuf.descriptor_pb2 import FieldDescriptorProto
+
+from ni_measurement_plugin_sdk_service._internal.parameter.decoder_strategy import get_type_default
+from ni_measurement_plugin_sdk_service._internal.parameter.metadata import (
+    ParameterMetadata,
+)
+from tests.utilities.stubs.loopback.types_pb2 import ProtobufColor
+
+
+def serialize_parameters(
+    parameter_metadata_dict: Dict[int, ParameterMetadata],
+    parameter_values: Sequence[Any],
+) -> bytes:
+    """Serialize the parameter values in same order based on the metadata_dict.
+
+    Args:
+        parameter_metadata_dict (Dict[int, ParameterMetadata]): Parameter metadata by ID.
+
+        parameter_value (Sequence[Any]): Parameter values to serialize.
+
+    Returns:
+        bytes: Serialized byte string containing parameter values.
+    """
+    pool = descriptor_pool.Default()
+    file_descriptor_proto = descriptor_pb2.FileDescriptorProto()
+    original_guid = str(uuid4())
+    unique_descriptor_name = str("msg" + "".join(filter(str.isalnum, original_guid))[:16])
+    file_descriptor_proto.name = unique_descriptor_name
+    file_descriptor_proto.package = unique_descriptor_name
+
+    # Create a DescriptorProto for the message
+    message_proto = file_descriptor_proto.message_type.add()
```

Comment:
Me and Joel talked and were confused how Python wouldn't be able throw the descriptor proto in the garbage collector since it wouldn't be used again after calling the function.

Reply:
The descriptor pool is a global object: `pool = descriptor_pool.Default()`. There is one default descriptor pool per process. `descriptor_pool.Default()` is an accessor function that returns the default descriptor pool: https://github.com/protocolbuffers/protobuf/blob/main/python/google/protobuf/descriptor_pool.py#L1354

Every time you call this function, you add a new descriptor to the pool. If you call it 500 times, the pool will contain 500 descriptors. If you call it 1,000,000 times, the pool will contain 1,000,000 descriptors.

Reply:
Put differently: you are creating a new message type every time you call this function. If you call it multiple times for the same service, you are creating unnecessary message types and storing them in the descriptor pool so they are never garbage collected.

If we passed more information to this function, then it could make a decision about whether to create a new message type or reuse a message type that it already created. Then it would only create a new message type the first time it's called for a specific service.

Reply:
Is it ok to to have one file descriptor proto then one message type per service in the pool instead of creating one file descriptor then adding the message types to it? It's not possible to update the file descriptor, that's already in the pool, with a new message type. Also, me and Joel couldn't find a way to remove a file descriptor from the pool to see if we could make a new, updated file descriptor then add that to the pool while removing the old one. So this was the best method Joel came up with.

Reply:
Creating a separate file descriptor per service is fine. Making an infinite number of calls to those services would consume a finite amount of memory, given a finite set of services.

My concern was that creating a file descriptor for each call consumes an infinite amount of memory, given a finite set of services.

Reply:
I think you should put both the Configurations and Outputs messages in the same file descriptor so they can share enums.